db_test.go

```diff
@@ -66,3 +69,160 @@ func TestLastStatus(t *testing.T) {
+	}
+}
+
+func TestNeedsUpgrade(t *testing.T) {
+	db, err := Open(dbPath, DBReadOnly)
+	if err != nil {
+		t.Fatalf("Open(%q): unexpected error: %s", dbPath, err)
+	}
+	defer db.Close()
+	if want, got := false, db.NeedsUpgrade(); want != got {
+		t.Errorf("db.NeedsUpgrade(): want %b got %b", want, got)
+	}
+}
+
+func TestUpgrade(t *testing.T) {
+	db, err := Open(dbPath, DBReadOnly)
+	if err != nil {
+		t.Fatalf("Open(%q): unexpected error: %s", dbPath, err)
+	}
+	defer db.Close()
+	if want, got := ErrReadOnlyDB, db.Upgrade(); want != got {
+		t.Errorf("db.Upgrade(): want error %q got %q", want, got)
+	}
+
+	db, err = Open(dbPath, DBReadWrite)
+	if err != nil {
+		t.Fatalf("Open(%q): unexpected error: %s", dbPath, err)
+	}
+	defer db.Close()
+	if want, got := error(nil), db.Upgrade(); want != got {
+		t.Errorf("db.Upgrade(): want error %q got %q", want, got)
+	}
+}
+
+func TestAddMessage(t *testing.T) {
+	fp, err := filepath.Abs("fixtures/emails/notmuch0202:2,")
+	if err != nil {
+		t.Fatalf("error getting the absolute path: %s", err)
+	}
+	nfp, err := filepath.Abs("fixtures/database-v1/new/notmuch0202:2,")
+	if err != nil {
+		t.Fatalf("error getting the absolute path: %s", err)
+	}
+
+	f, err := os.Open(fp)
+	if err != nil {
+		t.Fatalf("error opening the new email: %s", err)
+	}
+	defer f.Close()
+	nf, err := os.Create(nfp)
+	if err != nil {
+		t.Fatalf("error creating the new email: %s", err)
+	}
+	defer nf.Close()
+	if _, err := io.Copy(nf, f); err != nil {
+		t.Fatalf("error copying the email: %s", err)
+	}
+	defer os.Remove(nfp)
+
+	db, err := Open(dbPath, DBReadWrite)
+	if err != nil {
+		t.Fatalf("Open(%q): unexpected error: %s", dbPath, err)
+	}
+	defer db.Close()
+	msg, err := db.AddMessage(nfp)
+	if err != nil {
+		t.Fatalf("AddMessage(%q): got error: %s", nfp, err)
+	}
+	defer db.RemoveMessage(nfp)
+	if msg == nil {
+		t.Errorf("expecting msg to not be nil")
+	}
+}
+
+func TestFindMessage(t *testing.T) {
+	db, err := Open(dbPath, DBReadOnly)
+	if err != nil {
+		t.Fatalf("Open(%q): unexpected error: %s", dbPath, err)
+	}
+	defer db.Close()
+	if _, err := db.FindMessage("notfound"); err != ErrNotFound {
+		t.Errorf("db.FindMessage(\"notfound\"): expecting ErrNotFound got %s", err)
+	}
+	id := "87iqd9rn3l.fsf@vertex.dottedmag"
+	msg, err := db.FindMessage(id)
+	if err != nil {
+		t.Fatalf("db.FindMessage(%q): unexpected error: %s", id, err)
+	}
+	if want, got := id, msg.GetID(); want != got {
+		t.Errorf("db.FindMessage(%q).GetID(): want %s got %s", id, want, got)
+	}
+}
+
+func TestCompact(t *testing.T) {
+	backup := fmt.Sprintf("%s.backup", dbPath)
+	defer os.RemoveAll(backup)
+	if err := Compact(dbPath, backup); err != nil {
+		t.Fatalf("error compacting %q: %s", dbPath, err)
+	}
+
+	db, err := Open(dbPath, DBReadOnly)
+	if err != nil {
+		t.Fatalf("Open(%q): unexpected error: %s", dbPath, err)
+	}
+	defer db.Close()
+
+	// same test as TestFindMessage to ensure the database integrity.
```

Comment:
I usually don't do that in testing as I'd like the test to fully tell the story but I'm open to debate :)